rfcs/063-catalogue_pipeline_lambdas/README.md

```diff
@@ -0,0 +1,139 @@
+   works updates published to `<live_catalogue_pipeline>_ingestor_works_output` SNS topic. The ingestor service only runs when needed, eg. when we want to refresh the `authoritative-concepts` index.  
+
+   This allows for most efficient use of the infrastructure/resources and avoids one of the oft-cited complexity/downside in the catalogue-pipeline, namely that the application tries to handle two quite different use cases.  
+
+   The different versions of the aggregator and recorder are also configured for their purpose, thus making the infrastructure code easier to read and understand.
+
+## Suggested solution
+
+We could incrementally refactor the catalogue-pipeline services to use lambdas instead of ECS. Some considerations are:
+
+ - We currently build container images to packages our code for deployment, we may wish to continue doing so for simplicity. See the documentation around [packaging and deploying container based Java Lambdas](https://docs.aws.amazon.com/lambda/latest/dg/java-image.html) that we would need to use to achieve this.
+
+- Lambda has built-in scaling capabilities which will allow us to simplify the infrastructure by removing scaling configurations and Cloudwatch alarms on the queues. We would still use the `reindexing_state` to adjust data store provisioning. We 
+  could also leverage SQS lambda features to protect the data stores (eg. elasticsearch index at the ingestor stage) from overloading by setting a maximum concurrency on the event source.
+   
+- We can leverage SQS event source and lambda features to batch up to 10000 messages before invoking the function (useful in the context of the batcher service). This will make it possible to remove the batching mechanism that exists inside the 
+  services, making it easier to run them locally using a [lambda runtime emulator](https://github.com/aws/aws-lambda-runtime-interface-emulator).
+
+  Error handling could use [Partial batch responses](https://docs.aws.amazon.com/lambda/latest/dg/services-sqs-errorhandling.html#services-sqs-batchfailurereporting).
+  
+- Lambda is generally considered to be cost-effective for workloads with intermittent or unpredictable traffic patterns (1ms for 1024MB: $0.0000000167).
+
+- Processing speed could be slowed down by [AWS Lambda cold starts](https://aws.amazon.com/blogs/compute/operating-lambda-performance-optimization-part-1/) but this can be optimised by provisioning concurrency.
+  
+- The concepts-pipeline has demonstrated the value of using AWS Lambdas in a data pipeline.
+
+
+### Transition from ECS to Lambda
+
+In terms of transition from ECS to lambdas to prevent downtime and allow a smooth transition:
+
+- Update deployment to allow deployment of existing container images to both ECS & Lambda runtime environments.
+- Move services over one at a time, testing that a new Lambda service works by dual-running it alongside its ECS counterpart without publishing messages downstream.
```

Comment:
We'll need to do something to ensure that messages can be pulled by both systems.  Are we going to have a duplicate input queue populated by the same upstream SNS?

For stages that store data, rather than Batcher/Router/Matcher that just push messages downstream, do we want to consider in this RFC  how that is to be done? (just write to S3, regardless of whether it's a stage that needs to output an index?)

Reply:
> Are we going to have a duplicate input queue populated by the same upstream SNS?

Yes, that's what I imagined we'd do. Where services are stateful we'll have to take a decisions on how to handle state.